## Week 6 Homework Assignment

### Set up:
```
mkdir week6data
cd week6data
conda activate bioinfo
```
This code creates a file to store all work for this assignment, as well as sets the environment for the provided Makefile.

### Download the Provided Makefile:
A makefile is provided with this assignment. The use of this makefile is to create a signle pipeline file to download ```.fasta```, ```.gff```, and ```.fastq``` files, creates a ```bwa``` index using the provided ```.fasta``` file, and aligns the ```.fastq``` file to the index to generate a ```.bam``` file for alignment visualization.


### Makefile Options:

#### get_genome
```
get_genome genome=(NCBI genome accession number)
```
Using a provided NCBI genome accession number, downloads the ```.fasta``` and ```.gff``` file for the associated accession number, unzips the file, and moves these files to the current directory for ease of use with a name containing the accession number.

#### get_fastq
```
get_fastq fastq=(SRA accession number) genome_size=(genome size in base pairs) coverage=(number of reads per base pair) read_length=(length of a read in your desired SRA accession)
```
After being provided the genome size, desired coverage, and length of a read, downloads a set number of reads from a provided SRA accession number, provides a list of basic statistics, and runs FASTQC to assess the quality of the reads.

#### index
```
index genome=(NCBI genome accession number)
```
Grabs the provided ```.fasta``` file and creates bwa indexes for use during alignment.

#### align
```
align fastq=(SRA accession number)
```
Aligns the targeted ```.fastq``` file to the indexes within the work file to generate a ```.bam``` file alignment for visualization. Basic statistics are also calculated to determine the quality of the alignment.


### bam File Visualization for Zaire Ebolavirus (example)

#### 1. Download the Refseq for Zaire Ebolavirus
```
make get_genome genome=GCF_000848505.1
```

#### 2. Download a number of reads to get 10x genome coverage of reads from an assoicated SRR accession file
```
make get_fastq genome_size=18959 coverage=10 read_length=202 fastq=SRR1553500
```

#### 3. Index the genome
```
make index genome=GCF_000848505.1
```

#### 4. Align the Fastq file to the index to generate a bam file
```
make align fastq=SRR1553500
```

#### 5. Alignment of the .bam file to the Zaire Ebolavirus genome using IGV

![Zaire Ebolavirus IGV Alignment Overview](https://github.com/coledcaron/BMMB852Work/blob/main/Week_06/bam_igv_alignment.jpg)


### Question: What percentage of reads aligned to the genome?

According to the ```.txt``` file generated by the ```align``` rule, the number of aligned reads is:
```
reads mapped:	1842
reads mapped and paired:	1842	# paired-end technology bit set + both mates mapped
reads unmapped:	34
```
When the percentage is calculated, this means that 98.2% of the reads mapped to the provided genome.

### Question: What was the expected average coverage?

The initial coverage desired was a 10x genome coverage across the Zaire ebolavirus genome. After the mapping, it would be expected to have a 9.82x genome coverage to account for unmapped reads.

### Question: What is the observed average coverage?

By my eye, the observed average coverage value accross the genome is around 8x coverage. A majority of sites have around 7-9 reads mapping to any given site, with only a few locations deviating significantly outside of this range.

### Question: How much does the coverage vary across the genome? (Provide a visual estimate.)

Overall, I think that while the range in coverage difference is high, the variance is not too bad. The lowest mapped sites have around 2 reads at those sites, while the highest read sites have 25 reads for that spot. However, outside of these few sections, variation stays consistent at the 7-10 reads range for the length of the genome.

Of note, many of the sections with low read counts are often followed by areas of high read count. This may indicate that there is some restriction in this site that biases reads to be shifted away from given sites to pool in nearby areas.
