# Week 11 Homework Assignment

## Download the Provided Makefile:

A makefile is provided with this assignment. The use of this makefile is to create a signle pipeline file to download ```.fasta```, ```.gff```, and ```.fastq``` files, creates a ```bwa``` index using the provided ```.fasta``` file, and aligns the ```.fastq``` file to the index to generate a ```.bam``` file. From this file, a ```.bw``` and ```.vcf.gz``` file are generated to visualize the coverage and variants associated with the sequence. Finally, all associated files are opened and displayed in IGV for visualization.

--------------------------------


## bam File Visualization for Zaire Ebolavirus (example)

### Requirement: Please start and run IGV on your device before running any code from the Makefile. Certain commands in the makefile look to have IGV open to receive commands. If IGV is not open, the code will fail to initialize and return an error to open IGV.

### Option 1: Use the all option to run everything in one command

####  Run the all function with each variable set
```
make all genome=GCF_000848505.1 fastq=SRR1553500 genome_size=18959 coverage=10 read_length=202 sample=ZEBV_illumina
```

#### Output IGV Visual

![IGV_Visualization_of_a_Single_Sample](https://github.com/coledcaron/BMMB852Work/blob/main/Week_10/igv_images/single_sample_igv.jpg)

This IGV output displays the ```.vcf.gz```, ```.bw```, ```.bam```, and ```.gff``` files associated with the sample in one view. This can then be used to easily identify variant between the reference genome and sequencing sample, as well as their relation to the features in the ```.gff```.

---------------------------------

### Option 2: Run each code option individually

#### 1. Download the Refseq for Zaire Ebolavirus
```
make get_genome genome=GCF_000848505.1 sample=ZEBV
```

#### 2. Download a number of reads to get 10x genome coverage of reads from an assoicated SRR accession file
```
make get_fastq genome_size=18959 coverage=10 read_length=202 fastq=SRR1553500 sample=ZEBV
```

#### 3. Index the genome
```
make index sample=ZEBV
```

#### 4. Align the Fastq file to the index to generate a bam file
```
make align fastq=SRR155350 sample=ZEBV
```

#### 5. Convert the bam file into a bigwig file
```
make bigwig sample=ZEBV
```

#### 6. Create a vcf.gz file to map the variants in the sample
```
make vcf sample=ZEBV
```

#### 7. Visualize the .gff, .bam, .bw, and .vcf.gz in IGV
```
make igv sample=ZEBV
```

##### Output:


![IGV_Visualization_of_a_Single_Sample](https://github.com/coledcaron/BMMB852Work/blob/main/Week_10/igv_images/single_sample_igv.jpg)

This is the same output as would be generated by running the ```all``` function.

----------------------------

## VCF Effect Prediction

To evaluate the effect of each variant found within the ```vcf.gz``` file, I decided that it would be simplest ti visually inspect the ```vcf.gz``` in IGV to identify any variants and what their possible effect should be.

For each identified variant, I would reference an amino acid chart to determine if the amino acid at that location would change due to the mutation. If the amino acid changed, I would record the change and what I would expect the effect size would be, along with what possible effect it should cause. I screened variants from 8500 bp to the end of the genome, and only in protein coding genes. The variants with nonsilent mutations are recoreded below.

Variants Table:

| Genome Position and Nucleotide Change | Mutation Type | Amino Acid Change | Effect Size | Effect |
|:---:|:---:|:---:|:---:|:---|
| 10979 T>A | missense | M > K | major | nonpolar to postive charge |
| 11815 G>A | missense | V > I | minor | small size increase |
| 11970 A>C | missense | E > D | minor | small size increase |
| 12169 A>C | missense | M > L | minor | smal size decrease |
| 12223 C>T | missense | A > T | minor | small size increase |
| 12355 A>G | missense | I > V | minor | small size decrease |
| 12591 G>A | missense | M > I | minor | small size change |
| 12618 A>C | missense | Q > H | moderate | polar to positive charge |
| 13654 G>A | missense | D > N | major | negative charge to polar |
| 13856 A>G | missense | D > G | major | negative charge to neutral, large size change |
| 15004 T>G | missense | F > V | minor | large size decrease |
| 15728 G>A | missense | S > N | minor | small size change |
| 15794 G>A | missense | R > Q | moderate | positive charge to polar |
| 16360 T>C | missense | Y > H | major | large nonpolar to medium positive charge |
| 16401 A>C | missense | Q > H | moderate | polar to positive charge |
| 16424 G>A | missense | S > N | minor | small size change |
| 16540 T>G | missense | Y > D | major | large nonpolar to small negative charge |
| 16546 G>A | missense | A > T | minor | small size increase |
| 16552 G>A | missense | D > N | major | positive charge to polar |
| 16597 G>A | missense | E > K | major | positive charge to negative charge |
| 16649 A>G | missense | N > S | minor | small size change |
| 16805 T>C | missense | F > S | moderate | large nonpolar to small polar |
| 17057 G>A | missense | S > K | moderate | polar to positive charge |
| 17336 A>G | missense | K > R | minor | small size decrease |
| 17410 T>C | missense | Y > H | major | nonpolar to positive charge |
| 17421 A>G | missense | I > V | minor | small size decrease |
| 17733 G>T | missense | E > D | minor | small size decrease |
| 17833 A>G | missense | I > V | minor | small size decrease |

During my screening, I could not find any nonsense mutations, which would terminate a protein early. These types of mutations are usually deleterious, so finding a successful ebola infection with a nonsense mutation would be unlikely. Many of the missense mutations I found would have little effect, converting molecules of similar properties at many of these locations. A few variants are more interesting, converting residues to different charges, which is more likely to have a large effect on the resulting protein secondary and teritiary structure.

Two variants are of more interest to me. These are the variants at 16552 and 16597, which convert an asparte and glutamate residue to a asparagine and lysine residue. This converts these two negatively charged residues to a polar and positive charge respectively, which would drastically alters the characteristic of these two sites. 

![IGV Visualization of sites 16552 and 16597](https://github.com/coledcaron/BMMB852Work/blob/main/Week_11/vcf_effect_visualization.jpg)
